students/luyao_xu/lesson05/mailroom3.py

```diff
@@ -0,0 +1,148 @@
+
+        user_input = input(
+            'Please enter a full name, type "list" to see the list of donors, type "back" to go back: ')
+        if user_input.lower() == 'list':
+            for name in donations:
+                print(name['name'])
+            continue
+
+        elif user_input.lower() == 'back':
+            # back to the main menu
+            return
+        # If the user types a name in the list, use it.
+        elif user_input.lower() in [d['name'].lower() for d in donations]:
+            for d in donations:
+                if d['name'] == user_input:
+                    donate(d)
+                    # exit for loop on name match
+                    break
+        else:
+            # If the user types a name not in the list, add that name to the data structure and use it
+            d = create_donor(user_input)
+            donations.append(donate(d))
+            break
+
+
+def sort_total_given(k):
+    return sum(k['donation amount'])
+
+
+def report_string(d):
+    result = '{:<26} | {:^11} | {:^8} | {:>12}\n'.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+    result += '-' * len(result) + '\n'
+    for i in d:
+        name = i['name']
+        total_donation = sum(i['donation amount'])
+        gift_number = len(i['donation amount'])
+        average_donation = total_donation / gift_number
+
+        result += '{:<27} ${:>11.2f} {:>12} ${:>13.2f}\n'.format(name, total_donation, gift_number, average_donation)
+    return result
+
+
+def create_report():
+    donations.sort(key=sort_total_given, reverse=True)
+    #  print a list of your donors, sorted by total historical donation amount
+    print(report_string(donations))
+
+
+def letter(donor):
+    letter = 'Dear {}, we want to thank you for your total donation amount of ${}. Have a nice day!'.format(
+        donor['name'], donate_sum(donor))
+    return letter
+
+
+def send_to_everyone():
+    for donor in donations:
+        filename = Path('./letters/' + donor['name'] + '.txt')
+        filename.parent.mkdir(parents=True, exist_ok=True)
+        with filename.open('w') as outfile:
+            outfile.write(letter(donor))
+
+
+def menu_selection(prompt, dispatch_dict):
+    while True:
+        response = input(prompt)
+        result = dispatch_dict[response]()
+        if result == "exit":
+            break
+
+
+# quote menu function
+def quit_program():
+    print('Bye.')
+    return "exit"
+
+
+if __name__ == "__main__":
+
+    main_prompt = ("Please choose one of the actions in menu:\n1. Send a Thank you.\n2. Create a report.\n"
+                   "3. Send thank you letter to everyone.\n4. Quit.\nEnter the number of the action you want to make: ")
+
+    main_dispatch = {'1': send_thank,
+                     '2': create_report,
+                     '3': send_to_everyone,
+                     '4': quit_program}
+    try:
+        menu_selection(main_prompt, main_dispatch)
+    except KeyError:
```

Comment:
the `KeyError` needs to be around line 123 - I would consolidate the code:
```
def quit_program():
    print('bye')
    sys.exit()

...

menu_dispatch = <...>
while True:
    response = input("...")
    try:
        dispatch_dict[response]()
    except KeyError:
        print("invalid")

```
   
* simplify the code by mapping quit feature to sys.exit, avoid checking return
* code is more dense and clear, exception is caught around the code that can throw that exception